image-registry/remove-registry-baremetal.md

```diff
@@ -0,0 +1,137 @@
+
+### Implementation Details/Notes/Constraints
+
+- When the image registry operator bootstraps itself, it will inspect the cluster infrastructure platform status.
+- If the platform status type is `baremetal`, the registry will mark itself `Removed`.
+- A separate installation process will be resposible for provisioning storage. For RHV and k8s-native infra, this is expected to be a PVC with the `RWX` access mode.
+- Telemeter should be aware if a cluster's registry has been removed:
+  - Reason codes in the ClusterOperator conditions (`Available` in particular).
+  - Prometheus alerts warning registry has been removed.
+
+### Risks and Mitigations
+
+#### Image registry is never installed
+
+Several components will likely report errors if a customer runs the installer as `baremetal` and does nothing to set up the registry.
+Potential candidates that will break:
+
+1. Builds which push to an imagestream tag will error at runtime.
+2. Samples operator imagestreams which import from `registry.redhat.io` will require a pull secret to be usable.
+3. Imagestreams imported with `--reference-policy=local` will not reference the internal registry as expected. `BuildConfig` and `DeploymentConfig` objects which do their own resolution of `ImageStreamTag` will break.
+
+Mitigations:
+
+1. Existing error messages from affected components.
+2. Image registry's ClusterOperator will provide `Reason="Removed"` if the registry remains removed.
+3. If removed, the image registry operator should fire an alert warning that:
+   1. The registry has been removed.
+   2. The following components may not work as expected:
+       1. `ImageStreamTags`
+       2. `BuildConfigs` which reference `ImageStreamTags`
+       3. `DeploymentConfigs` which reference `ImageStreamTags`
+
+## Design Details
+
+### Test Plan
+
+- Existing e2e-operator suite will need to verify that the registry behaves as expeced when `Removed`
+  - No registry components are installed.
+  - Reporting "green"/does not block upgrade.
+  - Fire an alert
+  - Registry should behave as expected when state is switched to `Managed`
+- RHV should have their own e2e/conformance suite to ensure the registry comes up (likely out of scope for this enhancement)
+- k8s-native infra should have their own e2e/conformance suite to ensure the registry comes up (likely out of scope for this enhancement)
+
+### Graduation Criteria
+
+### Upgrade / Downgrade Strategy
+
+- Upgrades should work if registry is `Removed`
+- Downgrade to 4.2.z should work if registry is `Removed` (may require bugs to be backported and upgrade graph to go through a z-stream first)
+
+### Version Skew Strategy
+
+## Implementation History
+
+## Drawbacks
+
+The internal registry has always been a core part of OpenShift.
+This enhancement furthers the case that it may be OK to remove the registry.
+
+## Alternatives
+
+Continue the current path of defauling storage to `EmptyDir` on baremetal, and bring the registry up as `Managed`.
```

Comment:
Agreed

Reply:
long term alternative to evaluate is optionally installed outside of CVO.

Reply:
What if we created a `Warning` alert to indicate that the Image Registry is operating over `emptyDir` and then bootstrap any "unsupported" platform with this backend storage?

Reply:
Right

Reply:
@smarterclayton I am not sure about what your "Right" mean here. Does it mean that we are going for this alternative?

Reply:
@derekwaynecarr the image registry is needed to pull samples images from the Red Hat registry. And these samples are often used by builds. So it's not that optional.

Reply:
> @derekwaynecarr the image registry is needed to pull samples images from the Red Hat registry. And these samples are often used by builds. So it's not that optional.

@derekwaynecarr wants the samples to be optional too, but you're right that the samples won't be pullable for most users w/o a registry present to proxy the creds for the pull to registry.redhat.io, so not having it introduces another stumbling block for people, certainly.

I think @smarterclayton's right was probably in response to @derekwaynecarr's comment that long term the registry is installed outside the CVO.

> What if we created a Warning alert to indicate that the Image Registry is operating over emptyDir and then bootstrap any "unsupported" platform with this backend storage?

We discussed this fairly extensively, possibly before you joined the team (I know @adambkaplan was involved).  The main issue is ensuring we do *not* throw this alert if the admin has clearly expressly intended to configure emptydir storage. Distinguishing that from the "we picked emptydir because we don't understand this platform" is the slightly fiddly bit.

Assuming we address that distinction satisfactorily, I do think it's reasonable to divide the bootstrap logic into these 3 cases:


1) supported platforms that provide storage:  bootstrap the registry to use that storage (allocating/creating storage where needed, e.g. s3 buckets)

2) supported platforms that do not provide storage:  bootstrap the registry as `removed`.  If the platform installer includes "post-install" logic that creates storage (e.g. RHHI), that logic can also reconfigure the registry to `managed` and configure storage for it.

3) unknown/unrecognized/unsupported platforms:  bootstrap the registry w/ emptydir storage.  This ensures that we can do things like run CI against these clusters (sample imagestreams will work) even if we haven't yet had time to add explicit storage logic to the operator.  Generating a warning alert on these platforms makes sense.

Reply:
I think we might set the storage as `emptyDir` on config's `Status` but not on config's `Spec` when we are using it automiatically. This would show when it is something created by the user(on this case `Spec` would also contain the `emptyDir`). It would come with a price though, we will generate some inconsistencies(like when we automatically configure `S3` and set in `Spec` and consequently on `Status`).

I would prefer to go by the safest **always Removed** approach. If we need it online for a CI then it is part of the latter to set it up, that being said I recognize i don't fully understand the consequences or a `Removed` image registry.

Reply:
> I think we might set the storage as emptyDir on config's Status but not on config's Spec when we are configuring it automatically.

so we'd leave the spec storage config completely empty?  our controller is designed to look at spec and turn spec into reality (while then copying the value to status), so what you've done if you do that is essentially said "empty dir is the default if no spec config is set".  I don't think we want that.

> If we need it online for a CI then it is part of the latter to set it up, that being said I recognize i don't fully understand the consequences or a Removed image registry.

the primary consequence of defaulting to "removed" is bootstrapping CI (ie if a team wants to create a CI job for a new platform that the registry operator doesn't understand yet... if that CI job is running tests that require the registry, they won't be able to make it pass until the registry operator adds explicit support.  i'd prefer to avoid that, which is what using emptydir instead accomplishes).

Reply:
> so we'd leave the spec storage config completely empty? our controller is designed to look at spec and turn spec into reality (while then copying the value to status), so what you've done if you do that is essentially said "empty dir is the default if no spec config is set". I don't think we want that.

What I meant is: if there is `emptyDir` on `Status` but not in the `Spec` it means that it was automatically set up by the operator(and this case would happen only on certain platforms, as you mentioned: "unknown/unrecognized/unsupported" platforms). On the other hand if there is `emptyDir` on `Spec` and consequently on `Status` then it was something done deliberately by the user. It is just a way of knowing who chose `emptyDir`, user(manual) or operator(automatic) so we could base the alert on this.

Reply:
> What I meant is: if there is emptyDir on Status but not in the Spec it means that it was automatically set up by the operator

yeah, i understood the concept, but i'm not a fan of the behavior it results in.  In general spec indicates intent and status indicates whether we achieved that intent.  You are introducing effectively an implicit default in spec which says "if spec is empty, the default storage is emptydir".

If a user (for whatever reason) deleted their storage config out of spec, the controller logic would presumably conclude that it should set the storage to emptydir.

As an alternative, perhaps you can annotate the config resource with an annotation that indicates "operator defaulted this to emptydir, so we need to alert".  You can clear the annotation when the user sets the storage to something real.  And if a user ever sets the storage to emptydir, the annotation would not be there, so you'd alert.

Reply:
> yeah, i understood the concept, but i'm not a fan of the behavior it results in. In general spec indicates intent and status indicates whether we achieved that intent. You are introducing effectively an implicit default in spec which says "if spec is empty, the default storage is emptydir".

Yeah, this would be part of the "inconsistencies" I have mentioned on my comment, not appropriate indeed.

> As an alternative, perhaps you can annotate the config resource with an annotation that indicates "operator defaulted this to emptydir, so we need to alert". You can clear the annotation when the user sets the storage to something real. And if a user ever sets the storage to emptydir, the annotation would not be there, so you'd alert.

This makes sense. There will be a caveat if the user decides that he indeed likes to use EmptyDir(after we automatically configured it to EmptyDir) and never sets it to something real, on that case the only way of turning the alert off would be by manually removing the annotation, no?

Reply:
> This makes sense. There will be a caveat if the user decides that he indeed likes to use EmptyDir(after we automatically configured it to EmptyDir) and never sets it to something real, on that case the only way of turning the alert off would be by manually removing the annotation, no?

yeah, that is the edge case.